handler.go

```diff
@@ -5,107 +5,80 @@
 package main
 
 import (
+	"errors"
 	"fmt"
-	"io/ioutil"
 	"net/http"
-	"os"
 	"path/filepath"
 	"time"
 )
 
-func notFound(w http.ResponseWriter, err error) {
-	errString := ""
-	if err != nil {
-		errString = err.Error()
-	}
-
-	http.Error(w, errString, http.StatusNotFound)
+func notFound(w http.ResponseWriter, errorMessage string) {
+	http.Error(w, errorMessage, http.StatusNotFound)
 }
 
-func badRequest(w http.ResponseWriter, err error) {
-	errString := ""
-	if err != nil {
-		errString = err.Error()
-	}
+func notFoundError(w http.ResponseWriter, err error) {
+	http.Error(w, err.Error(), http.StatusNotFound)
+}
 
-	http.Error(w, errString, http.StatusBadRequest)
+func badRequest(w http.ResponseWriter, errorMessage string) {
+	http.Error(w, errorMessage, http.StatusBadRequest)
 }
 
-func catchAll(publicPath string, cacheTime time.Duration) func(w http.ResponseWriter, r *http.Request) {
+func catchAll(public http.FileSystem, cacheTime time.Duration) func(w http.ResponseWriter, r *http.Request) {
+	fileServer := http.FileServer(public)
 	return func(w http.ResponseWriter, r *http.Request) {
-
-		path := r.RequestURI
-
-		file, err := os.Stat(publicPath + path)
-		if os.IsNotExist(err) {
-			notFound(w, fmt.Errorf("404 Page Not Found Error"))
-			return
-		}
-
-		// Handles if it's a directory or last char is a / (also a directory)
-		// It then opens index.json by default (if it exists)
-		if len(path) == 0 {
-			path = "/index.json"
-		} else if path[len(path)-1:] == "/" {
-			path = path + "index.json"
-		} else if file.IsDir() {
-			path = path + "/index.json"
-		}
-
-		file, err = os.Stat(publicPath + path)
-		if os.IsNotExist(err) {
-			notFound(w, fmt.Errorf("404 Page Not Found Error"))
-			return
-		}
-
-		data, err := ioutil.ReadFile(publicPath + path)
+		path, err := determineResourcePath(r, public)
 		if err != nil {
-			notFound(w, fmt.Errorf("404 Page Not Found Error"))
+			notFoundError(w, err)
 			return
 		}
 
 		cacheHeaders(w, cacheTime)
-		sendHeader(w, r)
-		w.Write(data)
+
+		r.URL.Path = path
+		fileServer.ServeHTTP(w, r)
 	}
 }
 
-func jsonHeader(w http.ResponseWriter) {
-	w.Header().Set("Content-Type", "application/json")
-}
+func determineResourcePath(r *http.Request, public http.FileSystem) (string, error) {
+	path := r.RequestURI
+
+	// Handles if it's a directory or last char is a / (also a directory)
+	// It then opens index.json by default (if it exists)
+	if len(path) == 0 || path == "/" {
+		path = "index.json"
+	} else if path[len(path)-1:] == "/" {
+		path = filepath.Join(path, "index.json")
+	} else {
+		f, err := public.Open(path)
+		if err != nil { // catch all errors, including "forbidden access"
+			return "", errors.New("404 Page Not Found Error")
+		}
+		defer f.Close()
 
-func sendHeader(w http.ResponseWriter, r *http.Request) {
-	extension := filepath.Ext(r.RequestURI)
+		stat, err := f.Stat()
+		if err != nil {
+			return "", errors.New("404 Page Not Found Error")
+		}
 
-	switch extension {
```

Comment:
How are these headers managed with `http.FileSystem`?

Reply:
https://golang.org/src/net/http/sniff.go?s=646:688#L11

Reply:
Did you by chance check if we still ship the same ones as before or if there is a diff?

Reply:
I randomly checked SVG and TGZ, and it looks correct.

Reply:
Could you check all of them just to make sure if we have changes, we know about it? Unfortunately we don't have any tests for this yet.

What does it do for the ones where we set `charset=UTF-8`?

Reply:
Let me add some unit tests first and then we can come back to the discussion.

Reply:
@ruflin 
I've just added a unit test that verifies content-types based on the automatic detection. I assume it bases on some RFC. There are slight differences, but to my mind it's not fine to leave it as is, unless we need the information about content-type for further processing.

Reply:
I'm not too worried about the content type changes as my assumption is that the ones that are not the same, should have been adjusted anyways. My only concern would be, that we break something in Kibana (I assume not).